Jetmet uncertainties

veelken opened the pull request.

arizzi commented:
is the build in standalone broken by this ? I see a lot of cmssw/scram specific stuff

arizzi commented on python/postprocessing/modules/btv/btagSFProducer.py:
```diff
@@ -2,7 +2,7 @@
 import os
```
the changes in btagSF Producer seem to mix with the other PR

arizzi commented on python/postprocessing/modules/jme/jetmetUncertainties.py:
```diff
@@ -0,0 +1,408 @@
+        else:
+            pairs[ptc] = None
+    return pairs
+#----------------------------------------------------------------------------------------------------
```
the above two function should be added in tools.py rather than here

arizzi commented on python/postprocessing/modules/jme/jetmetUncertainties.py:
```diff
@@ -0,0 +1,408 @@
+        return deltaR2(e1.eta, e1.phi, p1.eta, p1.phi)
+    de = e1 - e2
+    dp = deltaPhi(p1, p2)
+    return de*de + dp*dp
```
aren't the above functions provided in the tools.py already?
I would import them from there

arizzi commented on python/postprocessing/modules/jme/jetmetUncertainties.py:
```diff
@@ -0,0 +1,408 @@
+
+def square(x):
+    return x*x
+
```
??

arizzi commented on python/postprocessing/modules/jme/jetmetUncertainties.py:
```diff
@@ -0,0 +1,408 @@
+
+        # initialize random number generator
+        # (needed for jet pT smearing)
+        self.rnd = ROOT.TRandom3(12345)
```
we should not do the smearing with TRandom but rather with pt_reco-pt_gen amplification

arizzi replied:
... I would also factorize the smearing into another file

arizzi commented on python/postprocessing/modules/jme/jetmetUncertainties.py:
```diff
@@ -0,0 +1,408 @@
+        #--------------------------------------------------------------------------------------------
+
+        params_resolution = ROOT.PyJetParametersWrapper()
+        params_resolution.setJetPt(jet.pt)
```
do we need to instantiate this at each jet or can it be a data member

veelken commented:
Hi Andrea,

I have addressed your comments. I have updated the btagSFProducer.py file that is present in this branch to the version on PR #23 that you merged today. Please have a look if everything is fine now and then merge this branch. Thanks!

arizzi commented on python/postprocessing/modules/jme/jetSmearer.py:
```diff
@@ -0,0 +1,125 @@
+        # (the txt files were downloaded from https://github.com/cms-jet/JRDatabase/tree/master/textFiles/Spring16_25nsV10_MC )
+        self.jerInputFilePath = os.environ['CMSSW_BASE'] + "/src/PhysicsTools/NanoAODTools/data/jme/"
+        self.jerInputFileName = "Spring16_25nsV10_MC_PtResolution_AK4PFchs.txt" 
+        self.jerUncertaintyInputFileName = "Spring16_25nsV10_MC_SF_AK4PFchs.txt"
```
can we make this configurable.. at least the filenames..

veelken commented:
Done.
